src/masternodes/accountshistory.cpp

```diff
@@ -7,18 +7,93 @@
 #include <masternodes/masternodes.h>
 #include <key_io.h>
 
-#include <limits>
-
-/// @attention make sure that it does not overlap with those in masternodes.cpp/tokens.cpp/undos.cpp/accounts.cpp !!!
+/// @Note it's in own database
 const unsigned char CAccountsHistoryView::ByAccountHistoryKey::prefix = 'h';
 
 void CAccountsHistoryView::ForEachAccountHistory(std::function<bool(AccountHistoryKey const &, CLazySerialize<AccountHistoryValue>)> callback, AccountHistoryKey const & start)
 {
     ForEach<ByAccountHistoryKey, AccountHistoryKey, AccountHistoryValue>(callback, start);
 }
 
-Res CAccountsHistoryView::SetAccountHistory(const AccountHistoryKey& key, const AccountHistoryValue& value)
+Res CAccountsHistoryView::WriteAccountHistory(const AccountHistoryKey& key, const AccountHistoryValue& value)
 {
     WriteBy<ByAccountHistoryKey>(key, value);
     return Res::Ok();
 }
+
+Res CAccountsHistoryView::EraseAccountHistory(const AccountHistoryKey& key)
+{
+    EraseBy<ByAccountHistoryKey>(key);
+    return Res::Ok();
+}
+
+CAccountHistoryStorage::CAccountHistoryStorage(const fs::path& dbName, std::size_t cacheSize, bool fMemory, bool fWipe)
+    : CStorageView(new CStorageLevelDB(dbName, cacheSize, fMemory, fWipe))
+{
+}
+
+CAccountsHistoryWriter::CAccountsHistoryWriter(CCustomCSView & storage, uint32_t height, uint32_t txn, const uint256& txid, uint8_t type, CAccountsHistoryView* historyView)
+    : CStorageView(new CFlushableStorageKV(storage.GetRaw())), height(height), txn(txn), txid(txid), type(type), historyView(historyView)
+{
+}
+
+Res CAccountsHistoryWriter::AddBalance(CScript const & owner, CTokenAmount amount)
+{
+    auto res = CCustomCSView::AddBalance(owner, amount);
+    if (historyView && res.ok && amount.nValue != 0) {
+        diffs[owner][amount.nTokenId] += amount.nValue;
```

Comment:
Does it need to use SafeAdd?

Reply:
`CCustomCSView::AddBalance` do `SafeAdd` so res will be not ok.